File: apps/cli/src/commands/export.ts
import { canExport, exportBundle, XsdValidationError } from "@spc-up/core";
import { getDb } from "@spc-up/db";
import { mkdir } from "node:fs/promises";
import path from "node:path";

export async function runExport(opts: {
  uf: string;
  exercicio: string;
  out: string;
}): Promise<void> {
  const exercicio = Number.parseInt(opts.exercicio, 10);
  if (Number.isNaN(exercicio)) {
    throw new Error("--exercicio must be a number");
  }

  const outDir = path.resolve(opts.out);
  await mkdir(outDir, { recursive: true });

  const db = getDb();

  if (!(await canExport(db, opts.uf, exercicio))) {
    console.error(
      "Exportação bloqueada: existem pendências ou bloqueio_export para este UF/exercício.",
    );
    process.exitCode = 1;
    return;
  }

  try {
    const paths = await exportBundle(db, opts.uf, exercicio, outDir);
    for (const target of paths) {
      if (path.basename(target) === "validacao.json") {
        continue;
      }
      console.log(`→ ${target}`);
    }
    console.log("Exportação concluída.");
  } catch (error) {
    if (error instanceof XsdValidationError) {
      console.error("XML inválido contra XSD SPCA; exportação não publicada.");
      for (const [filename, errors] of Object.entries(error.errorsByFile)) {
        if (errors.length === 0) {
          continue;
        }
        console.error(`  ${filename}:`);
        for (const message of errors) {
          console.error(`    - ${message}`);
        }
      }
      process.exitCode = 1;
      return;
    }

    const message = error instanceof Error ? error.message : String(error);
    console.error(message);
    process.exitCode = 1;
  }
}

File: legacy/python/spc_up/api/routes/export.py
"""SPCA XML export route."""

from __future__ import annotations

import io
import json
import zipfile
from pathlib import Path

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from sqlalchemy import select
from sqlalchemy.orm import Session

from spc_up.api.deps import get_db
from spc_up.models.entities import DiretorioEstadual
from spc_up.services.export.aplicacao import build_aplicacao_xml
from spc_up.services.export.doacao import build_doacao_xml
from spc_up.services.export.guard import can_export
from spc_up.services.export.origem import build_origem_xml
from spc_up.services.export.validation import (
    XsdValidationError,
    require_valid_xsd,
    validate_spca_exports,
)
from spc_up.spca.validate import SchemaName

router = APIRouter(prefix="/api/export", tags=["export"])


@router.get("/{uf}/{exercicio}")
def export_spca_zip(
    uf: str,
    exercicio: int,
    session: Session = Depends(get_db),
) -> StreamingResponse:
    uf = uf.upper()
    if not can_export(session, uf, exercicio):
        raise HTTPException(
            403,
            detail="Exportação bloqueada: existem pendências ou bloqueio_export neste UF/exercício.",
        )

    diretorio = session.scalar(select(DiretorioEstadual).where(DiretorioEstadual.uf == uf))
    if diretorio is None:
        raise HTTPException(404, f"Diretório estadual não cadastrado para UF={uf}")

    cnpj = diretorio.cnpj_prestador
    paths: list[tuple[SchemaName, Path]] = [
        ("origem", build_origem_xml(session, uf, exercicio, cnpj)),
        ("aplicacao", build_aplicacao_xml(session, uf, exercicio, cnpj)),
        ("doacao", build_doacao_xml(session, uf, exercicio, cnpj)),
    ]

    try:
        validacao = validate_spca_exports(paths)
        require_valid_xsd(validacao)
    except XsdValidationError as exc:
        raise HTTPException(
            422,
            detail={
                "message": "XML inválido contra XSD SPCA; exportação não publicada.",
                "errors": exc.errors_by_file,
            },
        ) from exc

    buffer = io.BytesIO()
    with zipfile.ZipFile(buffer, "w", zipfile.ZIP_DEFLATED) as zf:
        for _, path in paths:
            zf.write(path, arcname=path.name)
        zf.writestr("validacao.json", json.dumps(validacao, ensure_ascii=False, indent=2))

    buffer.seek(0)
    filename = f"spca_{uf}_{exercicio}.zip"
    return StreamingResponse(
        buffer,
        media_type="application/zip",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )
